### DISCUSSION TITLE
xmake removes arguments even with `{force = true }`.

### DISCUSSION THREAD

**Discussion by davidchisnall** [xmake-io/xmake#4762] [Q&A]
### Xmake Version

v2.8.6+20231226

### Operating System Version and Architecture

macOS 14.3

### Describe Bug

Xmake does not contain an exhaustive list of the compiler flags that a compiler supports, but acts as if it does.  Many compiler flags are positional and mean different things in different places.  Adding the same unknown flag with xmake causes it to be removed.  This makes it impossible to pass LLVM flags through clang, because each needs to be prefixed with `-mllvm`.  For example:

```
add_cflags("-mllvm", "-cheri-stack-bounds-single-intrinsic-threshold", "-mllvm", "1", { force = true })
```

This results in the following in the clang command line: `-mllvm -cheri-stack-bounds-single-intrinsic-threshold 1`.  

### Expected Behavior

Ideally, xmake should *never* tamper with arguments passed to `add_cflags` or similar.  These are low-level interfaces for communicating directly with the compiler.  If xmake wants to provide higher-level abstractions (as it does for debug level, optimisation level, and so on) that map compiler flags to other things then that's fine, but it should never assume that it knows more than the user about the flags that the compiler accepts.  Given that this is not the behaviour, at least users who explicitly tell xmake that they know what they're doing via `force = true` should just have the arguments forwarded to the compiler and not modified.

### Project Configuration

`xmake.lua`

```lua
target("example")
  add_cflags("-mllvm", "-some-flag", "-mllvm", "-some-other-flag", { force = true })
  add_files("a.c")
```

`a.c` is an empty file.  Run `xmake -v` and see the command line contains `-mllvm -some-flag -some-other-flag`, not `-mllvm -some-flag -mllvm -some-other-flag`.  

### Additional Information and Error Logs

N/A.

**Comment by waruqi** [comment]
xmake will remove duplicate flags.

try 

```lua
    add_cflags({"-mllvm", "-cheri-stack-bounds-single-intrinsic-threshold"}, {"-mllvm", "1"}, {expand = false, force = true })
```

**Reply by davidchisnall** [reply]
This appears to work, but the default behaviour is *very* dangerous.

A lot of toolchain flags are positional.  Many of them need to be repeated and they must appear in the correct order.  Some need to be specified multiple times to affect other things.

The default behaviour should be that xmake passes the flags that I ask for to the toolchain.  At the very least, it would be nice to have an option in the toolchain definition (not once per xmake.lua file) to disable xmake's mangling.

**Reply by Shiffted** [reply]
`-mllvm` and `-cheri-stack-bounds-single-intrinsic-threshold` aren't two different flags `-cheri-stack-bounds-single-intrinsic-threshold` is an argument to `-mllvm`, you need to combine them `"-mllvm -cheri-stack-bounds-single-intrinsic-threshold"` or `"-mllvm=-cheri-stack-bounds-single-intrinsic-threshold"`. The same goes for flags like `-Xclang`.

**Reply by davidchisnall** [reply]
If you combine them as strings, they're passed to as a single argument (single argv entry) and so the compiler errors.

We're hitting this again with `-target`, where we have a rule that amends it.  

Is there a flag I can set in the toolchain definition to tell xmake that it doesn't know what the flags are and shouldn't assume it knows more than the person adding them?
